File: JS/jwt-auth/src/auth/auth.service.ts
import {
  BadRequestException,
  Injectable,
  UnauthorizedException,
} from '@nestjs/common';
import { SignupDto } from './dto/signup.dto';
import { FileService } from './file.service';
import * as argon2 from 'argon2';
import { JwtService } from '@nestjs/jwt';
import { ConfigService } from '@nestjs/config';
import { RefreshDTO } from './dto/refresh.dto';

@Injectable()
export class AuthService {
  constructor(
    private readonly fileService: FileService,
    private readonly jwtService: JwtService,
    private readonly configService: ConfigService,
  ) {}
  async signUp(body: SignupDto) {
    const users = await this.fileService.readFile();
    const exists = users?.some((u) => u.username === body.username);

    if (exists) {
      throw new BadRequestException(
        `username ${body.username} is already taken`,
      );
    }

    const hashedPassword = await argon2.hash(body.password);

    const res = await this.fileService.appendFile({
      ...body,
      password: hashedPassword,
    });

    return {
      username: res.username,
      message: 'Signup successful',
    };
  }

  async login(body: SignupDto) {
    const users = await this.fileService.readFile();
    const user = users?.find((u) => u.username === body.username);

    if (!user) {
      throw new UnauthorizedException('Invalid credentials');
    }

    const isMatch = await argon2.verify(user.password, body.password);

    if (!isMatch) {
      throw new UnauthorizedException('Invalid credentials');
    }

    const payload = { sub: user.username, username: user.username };
    const accessToken = await this.jwtService.signAsync(payload, {
      secret: this.configService.get<string>('JWT_SECRET'),
      expiresIn: '15m',
    });

    const refreshToken = await this.jwtService.signAsync(payload, {
      secret: this.configService.get<string>('JWT_REFRESH_SECRET'),
      expiresIn: '7d',
    });

    const hashedRefreshToken = await argon2.hash(refreshToken);
    await this.fileService.updateUser(user.username, {
      refreshToken: hashedRefreshToken,
    });

    return { message: 'Login successful', accessToken, refreshToken };
  }

  async refresh(body: RefreshDTO) {
    let decoded: {
      sub: string;
      username: string;
    };
    try {
      decoded = await this.jwtService.verifyAsync(body.refreshToken, {
        secret: this.configService.get<string>('JWT_REFRESH_SECRET'),
      });
    } catch {
      throw new UnauthorizedException('Invalid or expired refresh token');
    }

    const users = await this.fileService.readFile();
    const user = users?.find((u) => u.username === decoded.username);

    if (!user || !user.refreshToken) {
      throw new UnauthorizedException('Invalid refresh token');
    }

    const isValid = await argon2.verify(user.refreshToken, body.refreshToken);
    if (!isValid) throw new UnauthorizedException('Invalid refresh token');

    const payload = { sub: user.username, username: user.username };
    const accessToken = await this.jwtService.signAsync(payload, {
      secret: this.configService.get<string>('JWT_SECRET'),
      expiresIn: '15m',
    });

    const refreshToken = await this.jwtService.signAsync(payload, {
      secret: this.configService.get<string>('JWT_REFRESH_SECRET'),
      expiresIn: '7d',
    });

    const hashedRefreshToken = await argon2.hash(refreshToken);
    await this.fileService.updateUser(user.username, {
      refreshToken: hashedRefreshToken,
    });

    return { accessToken, refreshToken };
  }
}

File: Go/jwt-auth/internal/services/auth_service.go
package services

import (
	"errors"
	"fmt"
	"jwt-auth/internal/models"

	"golang.org/x/crypto/bcrypt"
)

type AuthService struct {
	fileService *FileService
	jwtService  *JwtServie
}

func NewAuthService(fileService *FileService, jwtService *JwtServie) *AuthService {
	return &AuthService{fileService: fileService, jwtService: jwtService}
}

func (s *AuthService) SignUp(username, password string) error {
	users, err := s.fileService.ReadUsers()

	if err != nil {
		return fmt.Errorf("failed to read users: %w", err)
	}

	for _, u := range users {
		if u.Username == username {
			return errors.New("username already exists")
		}
	}

	hashedPass, err := hashPassword(password)
	if err != nil {
		return fmt.Errorf("%w", err)
	}

	users = append(users, models.User{
		Username: username,
		Password: hashedPass,
	})

	if err := s.fileService.WriteUsers(users); err != nil {
		return fmt.Errorf("failed to signup user %w", err)
	}

	return nil
}

func hashPassword(password string) (string, error) {
	hashedPass, err := bcrypt.GenerateFromPassword([]byte(password), bcrypt.DefaultCost)

	if err != nil {
		return "", errors.New("failed to hash password")
	}

	return string(hashedPass), nil
}

func (s *AuthService) Login(username, password string) (string, string, error) {
	users, err := s.fileService.ReadUsers()
	if err != nil {
		return "", "", fmt.Errorf("failed to read users: %w", err)
	}

	// Find user and verify password
	var foundUser *models.User

	for _, u := range users {
		if u.Username == username {
			foundUser = &u
			break
		}
	}

	fmt.Println(*foundUser)

	if foundUser == nil {
		return "", "", errors.New("invalid credentials")
	}

	accessToken, err := s.jwtService.GenerateAccessTok(username)
	if err != nil {
		return "", "", fmt.Errorf("failed to generate accessToken: %w", err)
	}

	refreshToken, err := s.jwtService.GenerateRefreshTok(username)
	if err != nil {
		return "", "", fmt.Errorf("failed to generate refreshToken: %w", err)
	}

	fmt.Println("Logged in")
	return accessToken, refreshToken, nil
}
